### PR TITLE
Add firewalld support + iptables fixes

### PR DESCRIPTION
**Pull request by rdoorn**
Hello,

As discussed:
- Added the firewalld provider support - this will be automatically picked for Centos7 + RHEL 7
- Undid some of the changes for chef 12 (like the provides statement in the providers) for the chef 11 compliance. also not sure how this would distinct between RHEL versions in the future.
- Added ability to remove rules.
- Fixed the order of iptables rules - i noticed these changed before, is the iptables parameter order different on other distros? order is key to identifying the rule already exists.
- Changed iptables -S to a iptables-save alternative grep.

Kind Regards,
Ronald

### CODE REVIEW

**Comment by martinb3** [comment]
Hello! I'd really like to be sure we have tests in place for each new provider, but this doesn't appear to even add a test suite for newer centos where firewalld is used. I'm going to create another branch, pull in your work, and also add tests. I'm also going to work on chef 11 compatibility using [the recommended pattern here](https://github.com/chef-cookbooks/mysql/blob/master/libraries/z_provider_mapping.rb). Thanks!

**Comment by rdoorn** [comment]
Hi @martinb3, thanks for that! your right on that tests are needed, had not yet build a unit test for this, but i guess its very simular to what you now have for iptables, but i wasn't sure how to distinct the OS versions in there, as its only as of version 7 that both RHEL and Centos start  using firewalld. i'll take on your examples for future reference.
also thanks for the chef 11 compatibility, that link is very usefull

Kind Regards,
Ronald

File: libraries/resource_firewall_rule.rb [hunk 1]
```diff
@@ -2,10 +2,11 @@ class Chef
   class Resource::FirewallRule < Resource
     include Poise(Chef::Resource::Firewall)
     provides :firewall_rule
+    attr_accessor :raw
```
Comment (martinb3) [inline libraries/resource_firewall_rule.rb]: There's already a raw parameter on the resource.

File: libraries/provider_firewall_rule_iptables.rb [hunk 2]
```diff
@@ -92,7 +122,15 @@ def build_firewall_rule(type = nil)
         if [:pre, :post].include?(new_resource.direction)
           firewall_rule << '-t nat '
         end
+
+        # Iptables order of prameters is important here see example output below:
```
Comment (martinb3) [inline libraries/provider_firewall_rule_iptables.rb]: It seems the output order is determined by the input order, so I'm not really sure this is necessary. If it is, we should write a test enforces it & demonstrates that it is required (test will fail when we change order).

Comment (rdoorn) [inline libraries/provider_firewall_rule_iptables.rb]: Hi Martin, in my tests on rhel, the output is static, no matter what the order of your input

example:
iptables -A INPUT -p tcp -m tcp --sport 80 -m state --state NEW -i eth -d 127.0.0.2 -s 127.0.0.8 -j ACCEPT
results in a iptables-save output of:
-A INPUT -s 127.0.0.8/32 -d 127.0.0.2/32 -i eth -p tcp -m tcp --sport 80 -m state --state NEW -j ACCEPT
not sure if this is RHEL specific, but on rhel 6.5 this was the case.

File: libraries/resource_firewall_rule.rb (hunk 1 above)
Comment (rdoorn) [inline libraries/resource_firewall_rule.rb]: the parameter is defined there so we could write to this parameter using the zap provider (zap uses this to remove the raw rules which are not defined using chef), i've found alternatives to fix this in the zap provider though.

**Pull request merged**